### PR TITLE
Feature/spend ecdsa

### PR DESCRIPTION
**Pull request by this-kramer**
## Reason for this PR:

- Add ECDSA version of Spend protocol to incentive system
- Minor changes in Earn according to changes in paper

### CODE REVIEW

File: crypto/src/main/java/org/cryptimeleon/incentive/crypto/callback/IDsidBlacklistHandler.java [hunk 1]
```diff
@@ -0,0 +1,12 @@
+package org.cryptimeleon.incentive.crypto.callback;
+
+import org.cryptimeleon.math.structures.rings.zn.Zn;
+
+/**
+ * Blacklist spend requests with different hash while allowing re-requesting with same hash gamma.
+ */
+public interface IDsidBlacklistHandler {
+    boolean containsDsidWithDifferentGamma(Zn.ZnElement doubleSpendingId, Zn.ZnElement gamma);
```
Comment (this-kramer) [inline crypto/src/main/java/org/cryptimeleon/incentive/crypto/callback/IDsidBlacklistHandler.java:9]: TODO: documentation of interface

File: crypto/src/main/java/org/cryptimeleon/incentive/crypto/callback/IStoreBasketRedeemedHandler.java [hunk 2]
```diff
@@ -1,25 +1,30 @@
...
 
+import org.cryptimeleon.math.structures.rings.zn.Zn;
+
 import java.math.BigInteger;
 import java.util.UUID;
 
 /**
  * Interface for the logic around redeeming baskets and re-requesting with the same hash.
  */
 public interface IStoreBasketRedeemedHandler {
```
Comment (this-kramer) [inline crypto/src/main/java/org/cryptimeleon/incentive/crypto/callback/IStoreBasketRedeemedHandler.java:11]: TODO: documentation

File: crypto/src/main/java/org/cryptimeleon/incentive/crypto/callback/ITransactionDBHandler.java [hunk 3]
```diff
@@ -0,0 +1,13 @@
+package org.cryptimeleon.incentive.crypto.callback;
+
+import org.cryptimeleon.incentive.crypto.model.EarnRequestECDSA;
+import org.cryptimeleon.incentive.crypto.model.SpendTransactionData;
+
+/**
+ * Interface to support a lambda function as a callback for adding clearing / doublespending protection data to some storage.
+ */
+public interface ITransactionDBHandler {
```
Comment (this-kramer) [inline crypto/src/main/java/org/cryptimeleon/incentive/crypto/callback/ITransactionDBHandler.java:9]: TODO: Add documentation

File: crypto/src/main/java/org/cryptimeleon/incentive/crypto/model/Token.java [hunk 4]
```diff
@@ -23,7 +23,8 @@ public class Token implements Representable, UniqueByteRepresentable {
     private GroupElement commitment0; // the first part of the Pedersen commitment computed from the bases and the exponents, representing the actual token
 
     @Represented(restorer = "G1")
-    private GroupElement commitment1; // the second part of the Pedersen commitment computed from the bases and the exponents, representing the actual token
+    private GroupElement commitment1; // TODO remove this. Is just g1Generator!
```
Comment (this-kramer) [inline crypto/src/main/java/org/cryptimeleon/incentive/crypto/model/Token.java:26]: TODO: Do this!

**Review by JanBobolz — APPROVED** [review]
👍

**Pull request merged**